app/src/main/java/com/hedvig/app/feature/marketing/MarketingViewModel.kt

```diff
@@ -56,108 +55,61 @@ class MarketingViewModel(
-    safeLet((newState as? PickMarket)?.market, (newState as? PickMarket)?.language) { m, l ->
-      updateApplicationLanguageUseCase.invoke(m, l)
-    }
-  }
+    updateApplicationLanguageUseCase.invoke(market, market.defaultLanguage())
 
-  private fun updateMarketPickerState(transform: (oldState: PickMarket) -> PickMarket): ViewState {
-    val oldState = state.value as? PickMarket
-      ?: throw IllegalStateException("Cannot update market picker state when not on market picker")
-    val newState = transform(oldState)
-    val newStateWithLanguage = newState.copy(
-      language = languageFromState(newState),
-    )
-    return _state.updateAndGet {
-      newStateWithLanguage.copy(
-        isValid = isValid(newStateWithLanguage),
+    _state.update {
+      it.copy(
+        market = market,
+        language = market.defaultLanguage(),
       )
     }
   }
 
-  private fun isValid(state: PickMarket): Boolean {
-    return (state.market != null && state.language != null)
-  }
-
-  private fun languageFromState(state: PickMarket): Language? {
-    val market = state.market ?: return state.language
-    val language = state.language ?: return market.defaultLanguage()
-
-    if (!market.isCompatible(language)) {
-      return market.defaultLanguage()
+  fun setLanguage(language: Language) {
+    state.value.market?.let {
+      updateApplicationLanguageUseCase.invoke(it, language)
     }
 
-    return language
+    _state.update { it.copy(language = language) }
   }
 
   fun submitMarketAndLanguage() {
-    val currentState = state.value as? PickMarket ?: throw IllegalStateException("Can't submit, not in PickMarket")
-    if (!currentState.isValid) {
-      throw IllegalStateException("Can't submit, PickMarket is not valid")
-    }
-    val market = currentState.market ?: throw IllegalStateException("Can't submit, market is null")
-    val language = currentState.language ?: throw IllegalStateException("Can't submit, language is null")
-    updateMarketPickerState { it.copy(isLoading = true) }
     viewModelScope.launch {
-      val result = submitMarketAndLanguagePreferencesUseCase
-        .invoke(language, market)
-      when (result) {
+      val market = _state.value.market ?: throw IllegalArgumentException("Market null")
+      val language = _state.value.language ?: throw IllegalArgumentException("Language null")
+
+      when (submitMarketAndLanguagePreferencesUseCase.invoke(language, market)) {
         is Either.Left -> {
-          updateMarketPickerState { it.copy(isLoading = false) }
+          _state.update { it.copy(isLoading = false) }
         }
         is Either.Right -> {
           updateApplicationLanguageUseCase.invoke(market, language)
           featureManager.invalidateExperiments()
-          _state.value = MarketPicked.Loading
-          loadMarketPicked()
+          _state.update {
+            it.copy(
+              selectedMarket = market,
+              isLoading = false,
+            )
+          }
```

Comment:
This was setting the state to loading, then called`featureManager.getLoginMethod()` inside the `loadMarketPicked` each time and then set that as a result. 
Now we're only changing the loginMethod at init{} as I see it.
However I see that here when we've submitted a new market and langauge we're doing a `featureManager.invalidateExperiments()` which means that this may in fact return a new value from that point on.
Unless I am missing something we'd need to do the same thing as before again on the Either.Right case. What do you think?

Reply:
Good catch. Fixing